---
title: CheckM
displayed_sidebar: multiqcSidebar
description: >
  Estimates genome completeness and contamination based on the presence or absence of marker genes
---

<!--
~~~~~ DO NOT EDIT ~~~~~
This file is autogenerated from the MultiQC module python docstring.
Do not edit the markdown, it will be overwritten.

File path for the source of this content: multiqc/modules/checkm/checkm.py
~~~~~~~~~~~~~~~~~~~~~~~
-->

:::note
Estimates genome completeness and contamination based on the presence or absence of marker genes

[https://github.com/Ecogenomics/CheckM](https://github.com/Ecogenomics/CheckM)
:::

The module parses the output files generated by CheckM.
It will only parse an output file from `checkm lineage_wf`, `checkm taxonomy_wf`, and `checkm qa`.
The output file needs to be in format 1 (`-o 1`).
All statistics for all samples are saved to `multiqc_data/checkm-table.txt`.

    Tested with CheckM v1.2.1

### File search patterns

```yaml
checkm:
  - contents_re: ".*Bin Id(?:\t| {3,})Marker lineage(?:\t| {3,})# genomes(?:\t| {3,})#\
      \ markers(?:\t| {3,})# marker sets.*"
    num_lines: 10
```
