---
title: goleft indexcov
description: >
  Quickly estimate coverage from a whole-genome bam index, providing 16KB resolution
---

<!--
~~~~~ DO NOT EDIT ~~~~~
This file is autogenerated from the MultiQC module python docstring.
Do not edit the markdown, it will be overwritten.

File path for the source of this content: multiqc/modules/goleft_indexcov/goleft_indexcov.py
~~~~~~~~~~~~~~~~~~~~~~~
-->

:::note
Quickly estimate coverage from a whole-genome bam index, providing 16KB resolution

[https://github.com/brentp/goleft/tree/master/indexcov](https://github.com/brentp/goleft/tree/master/indexcov)
:::

This is useful as a quick QC to get coverage values across the genome.

The module uses the PED and ROC data files to create diagnostic plots of coverage per
sample, helping to identify sample gender and coverage issues.

By default, we attempt to only plot chromosomes using standard human-like naming
(chr1, chr2... chrX or 1, 2 ... X) but you can specify chromosomes for detailed
ROC plots for alternative naming schemes in your configuration with:

```yaml
goleft_indexcov_config:
  chromosomes:
    - I
    - II
    - III
```

The number of plotted chromosomes is limited to 50 by default, you can customise this with the following:

```yaml
goleft_indexcov_config:
  max_chroms: 80
```

### File search patterns

```yaml
goleft_indexcov/ped:
  fn: "*-indexcov.ped"
goleft_indexcov/roc:
  fn: "*-indexcov.roc"
```
