---
name: Hostile
urls: ["https://github.com/bede/hostile"]
summary: >
  Removes host sequences from short and long read (meta)genomes, from paired or unpaired fastq[.gz]
---

<!--
~~~~~ DO NOT EDIT ~~~~~
This file is autogenerated from the MultiQC module python docstring.
Do not edit the markdown, it will be overwritten.

File path for the source of this content: multiqc/modules/hostile/hostile.py
~~~~~~~~~~~~~~~~~~~~~~~
-->

Hostile write the log in JSON format. Which is being used to generate the report.

```log
$ hostile clean --fastq1 human_1_1.fastq.gz --fastq2 human_1_2.fastq.gz >log.json
INFO: Hostile version 1.0.0. Mode: paired short read (Bowtie2)
INFO: Found cached standard index human-t2t-hla
INFO: Cleaningâ€¦
INFO: Cleaning complete
```

## JSON output

```log.json
[
    {
        "version": "1.0.0",
        "aligner": "bowtie2",
        "index": "human-t2t-hla",
        "options": [],
        "fastq1_in_name": "human_1_1.fastq.gz",
        "fastq1_in_path": "/path/to/human_1_1.fastq.gz",
        "fastq1_out_name": "human_1_1.clean_1.fastq.gz",
        "fastq1_out_path": "/path/to/human_1_1.clean_1.fastq.gz",
        "reads_in": 2,
        "reads_out": 0,
        "reads_removed": 2,
        "reads_removed_proportion": 1.0,
        "fastq2_in_name": "human_1_2.fastq.gz",
        "fastq2_in_path": "/path/to/human_1_2.fastq.gz",
        "fastq2_out_name": "human_1_2.clean_2.fastq.gz",
        "fastq2_out_path": "/path/to/human_1_2.clean_2.fastq.gz"
    }
]
```

A barplot using the JSON reports from different samples. Plot will shows the number of reads classified
as host-reads vs cleaned-reads (non-host reads).

### File search patterns

```yaml
hostile:
  contents: '"reads_removed_proportion"'
  fn: "*.json"
  num_lines: 100
```
