---
name: MetaPhlAn
urls: ["https://github.com/biobakery/MetaPhlAn"]
summary: >
  Profiles the composition of microbial communities from metagenomic shotgun sequencing data
---

<!--
~~~~~ DO NOT EDIT ~~~~~
This file is autogenerated from the MultiQC module python docstring.
Do not edit the markdown, it will be overwritten.

File path for the source of this content: multiqc/modules/metaphlan/metaphlan.py
~~~~~~~~~~~~~~~~~~~~~~~
-->

The module supports outputs from MetaPhlAn, that look like the following:

```tsv
k__Bacteria 2       100.0
k__Bacteria|p__Firmicutes   2|1239  44.30422
k__Bacteria|p__Bacteroidetes        2|976   34.73101
```

A bar graph is generated that shows the relative abundance for each sample that
fall into the top-10 categories for each taxa rank. The top categories are calculated
by summing the relative abundances across all samples.

Any species under the Additional Species column are ignored when making the graphs.

The number of top categories to plot can be customized in the config file:

```yaml
metaphlan:
  top_n: 10
```

### File search patterns

```yaml
metaphlan:
  contents: "#clade_name\tNCBI_tax_id\trelative_abundance\t"
  fn: "*.txt"
```
