---
title: Nonpareil
displayed_sidebar: multiqcSidebar
description: >
  Estimates metagenomic coverage and sequence diversity
---

<!--
~~~~~ DO NOT EDIT ~~~~~
This file is autogenerated from the MultiQC module python docstring.
Do not edit the markdown, it will be overwritten.

File path for the source of this content: multiqc/modules/nonpareil/nonpareil.py
~~~~~~~~~~~~~~~~~~~~~~~
-->

:::note
Estimates metagenomic coverage and sequence diversity

[https://github.com/lmrodriguezr/nonpareil](https://github.com/lmrodriguezr/nonpareil)
:::

Nonpareil uses the redundancy of the reads in a metagenomic dataset to estimate
the average coverage and predict the amount of sequences that will be required
to achieve "nearly complete coverage", defined as ≥95% or ≥99% average coverage.

Since Nonpareil main output has no model information, it is necessary extract the `curves` object as a `JSON` file.
From version `v3.5.5` this can be done with an auxiliary `R` script, briefly:

```bash
NonpareilCurves.R --json out.json model.npo
```

#### Module config options

The module plots a line graph for each sample, with a tab panel to switch between only observed data, only models,
or both combined (model with a dashed line). It will use the colors specified in the JSON file by `nonpareil` and,
if some is missing use one from a MultiQC colour scheme (default: Paired) that can be defined with:

```yaml
nonpareil:
  plot_colours: Paired
```

### File search patterns

```yaml
nonpareil:
  - contents: LRstar
    fn: "*.json"
    max_filesize: 1048576
    num_lines: 50
```
