---
title: QualiMap
displayed_sidebar: multiqcSidebar
description: >
  Quality control of alignment data and its derivatives like feature counts
---

<!--
~~~~~ DO NOT EDIT ~~~~~
This file is autogenerated from the MultiQC module python docstring.
Do not edit the markdown, it will be overwritten.

File path for the source of this content: multiqc/modules/qualimap/qualimap.py
~~~~~~~~~~~~~~~~~~~~~~~
-->

:::note
Quality control of alignment data and its derivatives like feature counts

[http://qualimap.bioinfo.cipf.es/](http://qualimap.bioinfo.cipf.es/)
:::

The module supports the Qualimap commands `BamQC` and `RNASeq`.
Note that Qualimap must be run with the `-outdir` option as well as
`-outformat HTML` (which is on by default). MultiQC uses files
found within the `raw_data_qualimapReport` folder (as well as `genome_results.txt`).

Qualimap adds lots of columns to the General Statistics table. To avoid making the table
too wide and bloated, some of these are hidden by default (`Error Rate`, `M Aligned`, `M Total reads`).
You can override these defaults in your MultiQC config file - for example, to show
`Error Rate` by default and hide `Ins. size` by default, add the following:

```yaml
table_columns_visible:
  QualiMap:
    general_error_rate: True
    median_insert_size: False
```

See the [relevant section of the documentation](https://docs.seqera.io/multiqc/reports/customisation#hiding-columns) for more detail.

In addition to this, it's possible to customise which coverage thresholds calculated
by the Qualimap BamQC module _(default: 1, 5, 10, 30, 50)_ and which of these are hidden in the
General Statistics tablewhen the report loads _(default: all hidden except 30X)_.

To do this, add something like the following to your MultiQC config file:

```yaml
qualimap_config:
  general_stats_coverage:
    - 10
    - 20
    - 40
    - 200
    - 30000
  general_stats_coverage_hidden:
    - 10
    - 20
    - 200
```

### File search patterns

```yaml
qualimap/bamqc/coverage:
  fn: coverage_histogram.txt
qualimap/bamqc/gc_dist:
  fn: mapped_reads_gc-content_distribution.txt
qualimap/bamqc/genome_fraction:
  fn: genome_fraction_coverage.txt
qualimap/bamqc/genome_results:
  fn: genome_results.txt
qualimap/bamqc/insert_size:
  fn: insert_size_histogram.txt
qualimap/rnaseq/coverage:
  fn: coverage_profile_along_genes_(total).txt
qualimap/rnaseq/rnaseq_results:
  fn: rnaseq_qc_results.txt
```
