---
title: Sylph-tax
displayed_sidebar: multiqcSidebar
description: >
    <p>Taxonomic profiling of metagenomic reads.</p>
---

<!--
~~~~~ DO NOT EDIT ~~~~~
This file is autogenerated from the MultiQC module python docstring.
Do not edit the markdown, it will be overwritten.

File path for the source of this content: multiqc/modules/sylphtax/sylphtax.py
~~~~~~~~~~~~~~~~~~~~~~~
-->

:::note
<p>Taxonomic profiling of metagenomic reads.</p>

[https://sylph-docs.github.io/](https://sylph-docs.github.io/), [https://sylph-docs.github.io/sylph-tax/](https://sylph-docs.github.io/sylph-tax/)
:::

The module supports outputs from sylphtax, that look like the following:

```tsv
clade_name	relative_abundance	sequence_abundance	ANI (if strain-level)
d__Bacteria	100.00010000000002	99.99999999999999	NA
d__Bacteria|p__Bacillota	24.640800000000002	18.712699999999998	NA
d__Bacteria|p__Bacillota_A	47.333499999999994	52.5969	NA
```

A bar graph is generated that shows the relative abundance for each sample that
fall into the top-10 categories for each taxa rank. The top categories are calculated
by summing the relative abundances across all samples.

The number of top categories to plot can be customized in the config file:

```yaml
sylphtax:
  top_n: 10
```

### File search patterns

```yaml
sylphtax:
  fn: '*.sylphmpa'
```
    