---
title: VCFTools
description: >
  Program to analyse and reporting on VCF files
---

<!--
~~~~~ DO NOT EDIT ~~~~~
This file is autogenerated from the MultiQC module python docstring.
Do not edit the markdown, it will be overwritten.

File path for the source of this content: multiqc/modules/vcftools/vcftools.py
~~~~~~~~~~~~~~~~~~~~~~~
-->

:::note
Program to analyse and reporting on VCF files

[https://vcftools.github.io](https://vcftools.github.io)
:::

#### Important General Note

- Depending on the size and density of the variant data (vcf),
  some of the stat files generated by vcftools can be very large.
  If you find that some of your input files are missing, increase
  the [config.log_filesize_limit](https://docs.seqera.io/multiqc/usage/troubleshooting#big-log-files)
  so that the large file(s) will not be skipped by MultiQC. Note, however,
  that this might make MultiQC very slow!

This module parses the outputs from VCFTools' various commands:

#### Implemented

- `relatedness2`
  - Plots a heatmap of pairwise sample relatedness.
  - Not to be confused with the similarly-named command `relatedness`
- `TsTv-by-count`
  - Plots the transition to transversion ratio as a function of
    alternative allele count (using only bi-allelic SNPs).
- `TsTv-by-qual`
  - Plots the transition to transversion ratio as a function of SNP
    quality threshold (using only bi-allelic SNPs).
- `TsTv-summary`
  - Plots a bargraph of the summary counts of each type of transition and
    transversion SNPs.

#### To do

VCFTools has a number of outputs not yet supported in MultiQC which
would be good to add. Please check GitHub if you'd like these added
or (better still), would like to contribute!

- basic stats
- relatedness
- freq
- depth
- [Everything else](https://vcftools.github.io/man_latest.html)

### File search patterns

```yaml
vcftools/relatedness2:
  fn: "*.relatedness2"
vcftools/tstv_by_count:
  fn: "*.TsTv.count"
vcftools/tstv_by_qual:
  fn: "*.TsTv.qual"
vcftools/tstv_summary:
  fn: "*.TsTv.summary"
```
