---
name: Long Ranger
urls: ["https://support.10xgenomics.com/genome-exome/software/pipelines/latest/what-is-long-ranger"]
summary: >
  Sample demultiplexing, barcode processing, alignment, quality control, variant calling, phasing, and structural variant calling
---

<!--
~~~~~ DO NOT EDIT ~~~~~
This file is autogenerated from the MultiQC module python docstring.
Do not edit the markdown, it will be overwritten.

File path for the source of this content: multiqc/modules/longranger/longranger.py
~~~~~~~~~~~~~~~~~~~~~~~
-->

Currently supported Longranger pipelines:

- `wgs`
- `targeted`
- `align`

Usage:

```bash
longranger wgs --fastqs=/path/to/fastq --id=NA12878
multiqc /path/to/NA12878
```

This module will look for the files `_invocation` and `summary.csv` in the `NA12878` folder, i.e. the output folder of Longranger in this example. The file `summary.csv` is required. If the file `_invocation` is not found the sample will receive a generic name in the MultiQC report (`longranger#1`), instead of `NA12878` or whatever was given by the `--id` parameter.

### File search patterns

```yaml
longranger/invocation:
  contents: call PHASER_SVCALLER_CS(
  fn: _invocation
  max_filesize: 2048
longranger/summary:
  contents: longranger_version,instrument_ids,gems_detected,mean_dna_per_gem,bc_on_whitelist,bc_mean_qscore,n50_linked_reads_per_molecule
  fn: "*summary.csv"
  num_lines: 2
```
