---
name: SortMeRNA
urls: ["http://bioinfo.lifl.fr/RNA/sortmerna/"]
summary: >
  Program for filtering, mapping and OTU-picking NGS reads in metatranscriptomic and metagenomic data
---

<!--
~~~~~ DO NOT EDIT ~~~~~
This file is autogenerated from the MultiQC module python docstring.
Do not edit the markdown, it will be overwritten.

File path for the source of this content: multiqc/modules/sortmerna/sortmerna.py
~~~~~~~~~~~~~~~~~~~~~~~
-->

The core algorithm is based on approximate seeds and allows for fast and sensitive analyses of nucleotide sequences. The main application of SortMeRNA is filtering ribosomal RNA from metatranscriptomic data.

The module parses the log files, which are created when `SortMeRNA` is run with the `--log` option.

The default header in the 'General Statistics' table is '% rRNA'. Users can override this using the configuration option:

```yaml
sortmerna:
  tab_header: "My database hits"
```

### File search patterns

```yaml
sortmerna:
  contents: Minimal SW score based on E-value
```
