---
title: Supported Tools
description: Tools supported by MultiQC
displayed_sidebar: multiqcSidebar
---

MultiQC currently has modules to support 153 bioinformatics tools, listed below.

Click the tool name to go to the MultiQC documentation for that tool.

:::tip[Missing something?]
If you would like another tool to to be supported, please [open an issue](https://github.com/MultiQC/MultiQC/issues/new?labels=module%3A+new&template=module-request.yml).
:::

import MultiqcModules from "@site/src/components/MultiqcModules";

<MultiqcModules
  modules={[
    {
      id: "modules/adapterremoval",
      data: {
        name: "Adapter Removal",
        summary:
          "Removes adapter sequences, trims low quality bases from 3' ends, or merges overlapping pairs into consensus",
      },
    },
    {
      id: "modules/afterqc",
      data: {
        name: "AfterQC",
        summary: "Automatic filtering, trimming, error removing, and quality control for FastQ data",
      },
    },
    {
      id: "modules/anglerfish",
      data: {
        name: "Anglerfish",
        summary: "Quality controls Illumina libraries sequenced on Oxford Nanopore flowcells",
      },
    },
    {
      id: "modules/bakta",
      data: { name: "Bakta", summary: "Rapid & standardized annotation of bacterial genomes, MAGs & plasmids" },
    },
    {
      id: "modules/bamdst",
      data: { name: "Bamdst", summary: "Lightweight tool to stat the depth coverage of target regions of BAM file(s)" },
    },
    {
      id: "modules/bamtools",
      data: {
        name: "Bamtools",
        summary: "Provides both a programmer's API and an end-user's toolkit for handling BAM files",
      },
    },
    {
      id: "modules/bbduk",
      data: {
        name: "BBDuk",
        summary: "Common data-quality-related trimming, filtering, and masking operations with a kmer based approach",
      },
    },
    {
      id: "modules/bbmap",
      data: {
        name: "BBTools",
        summary: "Pre-processing, assembly, alignment, and statistics tools for DNA/RNA sequencing reads",
      },
    },
    {
      id: "modules/bcftools",
      data: { name: "Bcftools", summary: "Utilities for variant calling and manipulating VCFs and BCFs" },
    },
    {
      id: "modules/bcl2fastq",
      data: {
        name: "bcl2fastq",
        summary: "Demultiplexes data and converts BCL files to FASTQ file formats for downstream analysis",
      },
    },
    {
      id: "modules/bclconvert",
      data: {
        name: "BCL Convert",
        summary: "Demultiplexes data and converts BCL files to FASTQ file formats for downstream analysis",
      },
    },
    {
      id: "modules/biobambam2",
      data: { name: "biobambam2", summary: "Tools for early stage alignment file processing" },
    },
    {
      id: "modules/biobloomtools",
      data: {
        name: "BioBloom Tools",
        summary:
          "Assigns reads to different references using bloom filters. This is faster than alignment and can be used for contamination detection",
      },
    },
    {
      id: "modules/biscuit",
      data: {
        name: "BISCUIT",
        summary: "Maps bisulfite converted DNA sequence reads and determines cytosine methylation states",
      },
    },
    {
      id: "modules/bismark",
      data: {
        name: "Bismark",
        summary: "Maps bisulfite converted sequence reads and determine cytosine methylation states",
      },
    },
    { id: "modules/bowtie1", data: { name: "Bowtie 1", summary: "Ultrafast, memory-efficient short read aligner" } },
    {
      id: "modules/bowtie2",
      data: {
        name: "Bowtie 2 / HiSAT2",
        summary: "Results from both Bowtie 2 and HISAT2, tools for aligning reads against a reference genome",
      },
    },
    { id: "modules/busco", data: { name: "BUSCO", summary: "Assesses genome assembly and annotation completeness" } },
    {
      id: "modules/bustools",
      data: {
        name: "Bustools",
        summary:
          "Tools for BUS files - a file format for single-cell RNA-seq data designed to facilitate the development of modular workflows for data processing",
      },
    },
    {
      id: "modules/ccs",
      data: {
        name: "CCS",
        summary: "PacBio tool that generates highly accurate single-molecule consensus reads (HiFi Reads)",
      },
    },
    {
      id: "modules/cellranger",
      data: { name: "Cell Ranger", summary: "Analyzes single cell expression or VDJ data produced by 10X Genomics" },
    },
    {
      id: "modules/checkqc",
      data: { name: "CheckQC", summary: "Checks a set of quality criteria against an Illumina runfolder" },
    },
    {
      id: "modules/clipandmerge",
      data: { name: "ClipAndMerge", summary: "Adapter clipping and read merging for ancient DNA data" },
    },
    {
      id: "modules/clusterflow",
      data: { name: "Cluster Flow", summary: "Simple and flexible bioinformatics pipeline tool" },
    },
    {
      id: "modules/conpair",
      data: { name: "Conpair", summary: "Estimates concordance and contamination for tumor\u2013normal pairs" },
    },
    {
      id: "modules/cutadapt",
      data: {
        name: "Cutadapt",
        summary: "Finds and removes adapter sequences, primers, poly-A tails, and other types of unwanted sequences",
      },
    },
    {
      id: "modules/damageprofiler",
      data: { name: "DamageProfiler", summary: "DNA damage pattern retrieval for ancient DNA analysis" },
    },
    {
      id: "modules/dedup",
      data: { name: "DeDup", summary: "Improved Duplicate Removal for merged/collapsed reads in ancient DNA analysis" },
    },
    {
      id: "modules/deeptools",
      data: { name: "deepTools", summary: "Tools to process and analyze deep sequencing data" },
    },
    {
      id: "modules/diamond",
      data: {
        name: "DIAMOND",
        summary: "Sequence aligner for protein and translated DNA searches, a drop-in replacement for the NCBI BLAST",
      },
    },
    {
      id: "modules/disambiguate",
      data: {
        name: "Disambiguate",
        summary: "Disambiguate reads aligned to two different species (e.g. human and mouse)",
      },
    },
    {
      id: "modules/dragen",
      data: {
        name: "DRAGEN",
        summary: "Illumina Bio-IT Platform that uses FPGA for secondary analysis of sequencing data",
      },
    },
    {
      id: "modules/dragen_fastqc",
      data: {
        name: "DRAGEN-FastQC",
        summary: "Illumina Bio-IT Platform that uses FPGA for secondary analysis of sequencing data",
      },
    },
    {
      id: "modules/eigenstratdatabasetools",
      data: {
        name: "eigenstratdatabasetools",
        summary:
          "Tools to compare and manipulate the contents of EingenStrat databases, and to calculate SNP coverage statistics in such databases",
      },
    },
    {
      id: "modules/fastp",
      data: {
        name: "fastp",
        summary: "All-in-one FASTQ preprocessor (QC, adapters, trimming, filtering, splitting...)",
      },
    },
    {
      id: "modules/fastq_screen",
      data: {
        name: "FastQ Screen",
        summary:
          "Screens a library of sequences in FastQ format against a set of sequence databases to see if the composition of the library matches with what you expect",
      },
    },
    {
      id: "modules/fastqc",
      data: { name: "FastQC", summary: "Quality control tool for high throughput sequencing data" },
    },
    {
      id: "modules/featurecounts",
      data: {
        name: "featureCounts",
        summary:
          "Counts mapped reads for genomic features such as genes, exons, promoter, gene bodies, genomic bins and chromosomal locations",
      },
    },
    { id: "modules/fgbio", data: { name: "fgbio", summary: "Processing and evaluating data containing UMIs" } },
    { id: "modules/filtlong", data: { name: "Filtlong", summary: "Filters long reads by quality" } },
    {
      id: "modules/flash",
      data: { name: "FLASh", summary: "Merges paired-end reads from next-generation sequencing experiments" },
    },
    { id: "modules/flexbar", data: { name: "Flexbar", summary: "Barcode and adapter removal tool" } },
    {
      id: "modules/freyja",
      data: { name: "Freyja", summary: "Recovers relative lineage abundances from mixed SARS-CoV-2 samples" },
    },
    {
      id: "modules/ganon",
      data: {
        name: "Ganon",
        summary:
          "Metagenomics classification: quickly assigns sequence fragments to their closest reference among thousands of references via Interleaved Bloom Filters of k-mer/minimizers",
      },
    },
    {
      id: "modules/gatk",
      data: { name: "GATK", summary: "Wide variety of tools with a primary focus on variant discovery and genotyping" },
    },
    {
      id: "modules/gffcompare",
      data: {
        name: "GffCompare",
        summary: "Tool to compare, merge and annotate one or more GFF files with a reference annotation in GFF format",
      },
    },
    { id: "modules/glimpse", data: { name: "GLIMPSE", summary: "Low-coverage whole genome sequencing imputation" } },
    {
      id: "modules/goleft_indexcov",
      data: {
        name: "goleft indexcov",
        summary: "Quickly estimate coverage from a whole-genome bam index, providing 16KB resolution",
      },
    },
    { id: "modules/gopeaks", data: { name: "GoPeaks", summary: "Calls peaks in CUT&TAG/CUT&RUN datasets" } },
    {
      id: "modules/happy",
      data: { name: "hap.py", summary: "Benchmarks variant calls against gold standard truth datasets" },
    },
    {
      id: "modules/hicexplorer",
      data: { name: "HiCExplorer", summary: "Hi-C analysis from processing to visualization" },
    },
    { id: "modules/hicpro", data: { name: "HiC-Pro", summary: "Pipeline for Hi-C data processing" } },
    { id: "modules/hicup", data: { name: "HiCUP", summary: "Mapping and quality control on Hi-C data" } },
    {
      id: "modules/hifiasm",
      data: { name: "HiFiasm", summary: "Haplotype-resolved assembler for accurate Hifi reads" },
    },
    {
      id: "modules/hisat2",
      data: { name: "HISAT2", summary: "Maps DNA or RNA reads against a genome or a population of genomes" },
    },
    { id: "modules/homer", data: { name: "HOMER", summary: "Motif discovery and next-gen sequencing analysis" } },
    {
      id: "modules/hops",
      data: {
        name: "HOPS",
        summary: "Ancient DNA characteristics screening tool of output from the metagenomic aligner MALT",
      },
    },
    {
      id: "modules/hostile",
      data: {
        name: "Hostile",
        summary: "Removes host sequences from short and long read (meta)genomes, from paired or unpaired fastq[.gz]",
      },
    },
    {
      id: "modules/htseq",
      data: {
        name: "HTSeq Count",
        summary: "Part of the HTSeq package: counts reads covering specified genomic features",
      },
    },
    {
      id: "modules/humid",
      data: {
        name: "HUMID",
        summary: "Reference-free tool to quickly remove duplicates from FastQ files, with or without UMIs",
      },
    },
    {
      id: "modules/interop",
      data: { name: "Illumina InterOp Statistics", summary: "Reading and writing InterOp metric files" },
    },
    {
      id: "modules/isoseq",
      data: {
        name: "Iso-Seq",
        summary: "Identifies transcripts in PacBio single-molecule sequencing data (HiFi reads)",
      },
    },
    { id: "modules/ivar", data: { name: "iVar", summary: "Functions for viral amplicon-based sequencing" } },
    {
      id: "modules/jcvi",
      data: { name: "JCVI Genome Annotation", summary: "Computes statistics on genome annotation" },
    },
    { id: "modules/jellyfish", data: { name: "Jellyfish", summary: "Counting k-mers in DNA" } },
    { id: "modules/kaiju", data: { name: "Kaiju", summary: "Taxonomic classification for metagenomics" } },
    {
      id: "modules/kallisto",
      data: {
        name: "Kallisto",
        summary: "Quantifies abundances of transcripts (or more generally, of target sequences) from RNA-Seq data",
      },
    },
    {
      id: "modules/kat",
      data: { name: "K-mer Analysis Toolkit", summary: "Analyses sequencing data via its k-mer spectra" },
    },
    {
      id: "modules/kraken",
      data: {
        name: "Kraken",
        summary:
          "Taxonomic classification using exact k-mer matches to find the lowest common ancestor (LCA) of a given sequence",
      },
    },
    { id: "modules/leehom", data: { name: "leeHom", summary: "Bayesian reconstruction of ancient DNA" } },
    {
      id: "modules/librarian",
      data: {
        name: "Librarian",
        summary: "Predicts the sequencing library type from the base composition of a FastQ file",
      },
    },
    { id: "modules/lima", data: { name: "Lima", summary: "Demultiplex PacBio single-molecule sequencing reads" } },
    {
      id: "modules/longranger",
      data: {
        name: "Long Ranger",
        summary:
          "Sample demultiplexing, barcode processing, alignment, quality control, variant calling, phasing, and structural variant calling",
      },
    },
    {
      id: "modules/macs2",
      data: { name: "MACS2", summary: "Identifies transcription factor binding sites in ChIP-seq data" },
    },
    {
      id: "modules/malt",
      data: {
        name: "MALT",
        summary:
          "Aligns of metagenomic reads to a database of reference sequences (such as NR, GenBank or Silva) and outputs a MEGAN RMA file",
      },
    },
    {
      id: "modules/mapdamage",
      data: { name: "mapDamage", summary: "Tracks and quantifies damage patterns in ancient DNA sequences" },
    },
    { id: "modules/megahit", data: { name: "MEGAHIT", summary: "NGS read assembler" } },
    {
      id: "modules/metaphlan",
      data: {
        name: "MetaPhlAn",
        summary: "Profiles the composition of microbial communities from metagenomic shotgun sequencing data",
      },
    },
    {
      id: "modules/methylqa",
      data: { name: "methylQA", summary: "Methylation sequencing data quality assessment tool" },
    },
    {
      id: "modules/minionqc",
      data: { name: "MinIONQC", summary: "Quality control for ONT (Oxford Nanopore) long reads" },
    },
    {
      id: "modules/mirtop",
      data: {
        name: "mirtop",
        summary: "Annotates miRNAs and isomiRs and compute general statistics in mirGFF3 format",
      },
    },
    { id: "modules/mirtrace", data: { name: "miRTrace", summary: "Quality control for small RNA sequencing data" } },
    {
      id: "modules/mosdepth",
      data: { name: "Mosdepth", summary: "Fast BAM/CRAM depth calculation for WGS, exome, or targeted sequencing" },
    },
    {
      id: "modules/motus",
      data: {
        name: "Motus",
        summary: "Microbial profiling through marker gene (MG)-based operational taxonomic units (mOTUs)",
      },
    },
    {
      id: "modules/mtnucratio",
      data: { name: "mtnucratio", summary: "Computes mitochondrial to nuclear genome ratios in NGS datasets" },
    },
    {
      id: "modules/multivcfanalyzer",
      data: {
        name: "MultiVCFAnalyzer",
        summary:
          "Reads multiple VCF files into combined genotype calls, produces summary statistics and downstream formats",
      },
    },
    {
      id: "modules/nanoq",
      data: { name: "nanoq", summary: "Reports read quality and length from nanopore sequencing data" },
    },
    {
      id: "modules/nanostat",
      data: {
        name: "NanoStat",
        summary:
          "Reports various statistics for long read dataset in FASTQ, BAM, or albacore sequencing summary format (supports NanoPack; NanoPlot, NanoComp)",
      },
    },
    {
      id: "modules/nextclade",
      data: {
        name: "Nextclade",
        summary: "Viral genome alignment, clade assignment, mutation calling, and quality checks",
      },
    },
    { id: "modules/ngsbits", data: { name: "ngs-bits", summary: "Calculating statistics from FASTQ, BAM, and VCF" } },
    {
      id: "modules/ngsderive",
      data: {
        name: "ngsderive",
        summary: "Forensic tool for by backwards computing library information in sequencing data",
      },
    },
    {
      id: "modules/nonpareil",
      data: { name: "Nonpareil", summary: "Estimates metagenomic coverage and sequence diversity" },
    },
    {
      id: "modules/odgi",
      data: {
        name: "ODGI",
        summary: "Analysis and manipulation of pangenome graphs structured in the variation graph model",
      },
    },
    {
      id: "modules/optitype",
      data: { name: "OptiType", summary: "Precision HLA typing from next-generation sequencing data" },
    },
    {
      id: "modules/pairtools",
      data: {
        name: "pairtools",
        summary:
          "Toolkit for Chromatin Conformation Capture experiments. Handles short-reads paired reference alignments, extracts 3C-specific information, and perform common tasks such as sorting, filtering, and deduplication",
      },
    },
    {
      id: "modules/pangolin",
      data: {
        name: "Pangolin",
        summary: "Uses variant calls to assign SARS-CoV-2 genome sequences to global lineages",
      },
    },
    {
      id: "modules/pbmarkdup",
      data: {
        name: "pbmarkdup",
        summary:
          "Takes one or multiple sequencing chips of an amplified libray as HiFi reads and marks or removes duplicates",
      },
    },
    {
      id: "modules/peddy",
      data: {
        name: "Peddy",
        summary: "Compares familial-relationships and sexes as reported in a PED file with those inferred from a VCF",
      },
    },
    {
      id: "modules/phantompeakqualtools",
      data: {
        name: "phantompeakqualtools",
        summary: "Computes informative enrichment and quality measures for ChIP-seq/DNase-seq/FAIRE-seq/MNase-seq data",
      },
    },
    {
      id: "modules/picard",
      data: { name: "Picard", summary: "Tools for manipulating high-throughput sequencing data" },
    },
    {
      id: "modules/porechop",
      data: { name: "Porechop", summary: "Finds and removes adapters from Oxford Nanopore reads" },
    },
    {
      id: "modules/preseq",
      data: {
        name: "Preseq",
        summary:
          "Estimates library complexity, showing how many additional unique reads are sequenced for increasing total read count",
      },
    },
    {
      id: "modules/prinseqplusplus",
      data: {
        name: "PRINSEQ++",
        summary:
          "C++ implementation of the prinseq-lite.pl program. Filters, reformats, and trims genomic and metagenomic reads",
      },
    },
    { id: "modules/prokka", data: { name: "Prokka", summary: "Rapid annotation of prokaryotic genomes" } },
    {
      id: "modules/purple",
      data: { name: "PURPLE", summary: "A purity, ploidy and copy number estimator for whole genome tumor data" },
    },
    {
      id: "modules/pychopper",
      data: {
        name: "Pychopper",
        summary: "Identifies, orients, trims and rescues full length Nanopore cDNA reads. Can also rescue fused reads",
      },
    },
    {
      id: "modules/pycoqc",
      data: {
        name: "pycoQC",
        summary: "Computes metrics and generates interactive QC plots for Oxford Nanopore technologies sequencing data",
      },
    },
    {
      id: "modules/qc3C",
      data: { name: "qc3C", summary: "Reference-free and BAM based quality control for Hi-C data" },
    },
    {
      id: "modules/qorts",
      data: { name: "QoRTs", summary: "Toolkit for analysis, QC, and data management of RNA-Seq datasets" },
    },
    {
      id: "modules/qualimap",
      data: { name: "QualiMap", summary: "Quality control of alignment data and its derivatives like feature counts" },
    },
    { id: "modules/quast", data: { name: "QUAST", summary: "Quality assessment tool for genome assemblies" } },
    {
      id: "modules/rna_seqc",
      data: { name: "RNA-SeQC", summary: "RNA-Seq metrics for quality control and process optimization" },
    },
    {
      id: "modules/rockhopper",
      data: {
        name: "Rockhopper",
        summary: "Bacterial RNA-seq analysis: align reads to coding sequences, rRNAs, tRNAs, and miscellaneous RNAs",
      },
    },
    {
      id: "modules/rsem",
      data: { name: "RSEM", summary: "Estimates gene and isoform expression levels from RNA-Seq data" },
    },
    { id: "modules/rseqc", data: { name: "RSeQC", summary: "Evaluates high throughput RNA-seq data" } },
    {
      id: "modules/salmon",
      data: { name: "Salmon", summary: "Quantifies expression of transcripts using RNA-seq data" },
    },
    { id: "modules/sambamba", data: { name: "Sambamba", summary: "Toolkit for interacting with BAM/CRAM files" } },
    {
      id: "modules/samblaster",
      data: { name: "Samblaster", summary: "Marks duplicates and extracts discordant and split reads from sam files" },
    },
    { id: "modules/samtools", data: { name: "Samtools", summary: "Toolkit for interacting with BAM/CRAM files" } },
    {
      id: "modules/sargasso",
      data: { name: "Sargasso", summary: "Separates mixed-species RNA-seq reads according to their species of origin" },
    },
    {
      id: "modules/seqera_cli",
      data: { name: "Seqera Platform CLI", summary: "Reports statistics generated by the Seqera Platform CLI" },
    },
    {
      id: "modules/sequali",
      data: { name: "Sequali", summary: "Sequencing quality control for both long-read and short-read data" },
    },
    {
      id: "modules/seqwho",
      data: {
        name: "SeqWho",
        summary:
          "Determines FASTQ(A) sequencing file source protocol and the species of origin, to check that the composition of the library is expected",
      },
    },
    {
      id: "modules/seqyclean",
      data: { name: "SeqyClean", summary: "Filters adapters, vectors, and contaminants while quality trimming" },
    },
    {
      id: "modules/sexdeterrmine",
      data: {
        name: "SexDetErrmine",
        summary:
          "Calculates relative coverage of X and Y chromosomes and their associated error bars from the depth of coverage at specified SNPs",
      },
    },
    {
      id: "modules/sickle",
      data: { name: "Sickle", summary: "A windowed adaptive trimming tool for FASTQ files using quality" },
    },
    { id: "modules/skewer", data: { name: "Skewer", summary: "Adapter trimming tool for NGS paired-end sequences" } },
    { id: "modules/slamdunk", data: { name: "Slamdunk", summary: "Tool to analyze SLAM-Seq data" } },
    {
      id: "modules/snippy",
      data: { name: "Snippy", summary: "Rapid haploid variant calling and core genome alignment" },
    },
    {
      id: "modules/snpeff",
      data: {
        name: "SnpEff",
        summary: "Annotates and predicts the effects of variants on genes (such as amino acid changes)",
      },
    },
    {
      id: "modules/snpsplit",
      data: {
        name: "SNPsplit",
        summary: "Allele-specific alignment sorter. Determines allelic origin of reads that cover known SNP positions",
      },
    },
    {
      id: "modules/somalier",
      data: {
        name: "Somalier",
        summary: "Genotype to pedigree correspondence checks from sketches derived from BAM/CRAM or VCF",
      },
    },
    {
      id: "modules/sortmerna",
      data: {
        name: "SortMeRNA",
        summary: "Program for filtering, mapping and OTU-picking NGS reads in metatranscriptomic and metagenomic data",
      },
    },
    {
      id: "modules/sourmash",
      data: { name: "Sourmash", summary: "Quickly searches, compares, and analyzes genomic and metagenomic data sets" },
    },
    {
      id: "modules/spaceranger",
      data: { name: "Space Ranger", summary: "Tool to analyze 10x Genomics spatial transcriptomics data" },
    },
    {
      id: "modules/stacks",
      data: { name: "Stacks", summary: "Analyzes restriction enzyme-based data (e.g. RAD-seq)" },
    },
    { id: "modules/star", data: { name: "STAR", summary: "Universal RNA-seq aligner" } },
    {
      id: "modules/supernova",
      data: { name: "Supernova", summary: "De novo genome assembler of 10X Genomics linked-reads" },
    },
    {
      id: "modules/theta2",
      data: { name: "THetA2", summary: "Estimates tumour purity and clonal / subclonal copy number" },
    },
    {
      id: "modules/tophat",
      data: { name: "Tophat", summary: "Splice junction RNA-Seq reads mapper for mammalian-sized genomes" },
    },
    { id: "modules/trimmomatic", data: { name: "Trimmomatic", summary: "Read trimming tool for Illumina NGS data" } },
    {
      id: "modules/truvari",
      data: { name: "Truvari", summary: "Benchmarking, merging, and annotating structural variants" },
    },
    {
      id: "modules/umicollapse",
      data: {
        name: "UMICollapse",
        summary: "Algorithms for efficiently collapsing reads with Unique Molecular Identifiers",
      },
    },
    {
      id: "modules/umitools",
      data: {
        name: "UMI-tools",
        summary: "Tools for dealing with Unique Molecular Identifiers (UMIs)/(RMTs) and scRNA-Seq barcodes",
      },
    },
    {
      id: "modules/varscan2",
      data: { name: "VarScan2", summary: "Variant detection in massively parallel sequencing data" },
    },
    { id: "modules/vcftools", data: { name: "VCFTools", summary: "Program to analyse and reporting on VCF files" } },
    {
      id: "modules/vep",
      data: {
        name: "VEP",
        summary:
          "Determines the effect of variants on genes, transcripts and protein sequences, as well as regulatory regions",
      },
    },
    {
      id: "modules/verifybamid",
      data: { name: "VerifyBAMID", summary: "Detects sample contamination and/or sample swaps" },
    },
    {
      id: "modules/vg",
      data: { name: "VG", summary: "Toolkit to manipulate and analyze graphical genomes, including read alignment" },
    },
    {
      id: "modules/whatshap",
      data: {
        name: "WhatsHap",
        summary: "Phasing genomic variants using DNA reads (aka read-based phasing, or haplotype assembly)",
      },
    },
    {
      id: "modules/xengsort",
      data: { name: "Xengsort", summary: "Fast xenograft read sorter based on space-efficient k-mer hashing" },
    },
    { id: "modules/xenome", data: { name: "Xenome", summary: "Classifies reads from xenograft sources" } },
  ]}
/>
