---
title: Supported Tools
description: Tools supported by MultiQC
displayed_sidebar: multiqcSidebar
---

MultiQC currently has modules to support 153 bioinformatics tools, listed below.

Click the tool name to go to the MultiQC documentation for that tool.

:::tip[Missing something?]
If you would like another tool to to be supported, please [open an issue](https://github.com/MultiQC/MultiQC/issues/new?labels=module%3A+new&template=module-request.yml).
:::

| Tool                                                          | Description                                                                                                                                                                                                       |
| ------------------------------------------------------------- | ----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| [Adapter Removal](modules/adapterremoval.md)                  | Removes adapter sequences, trims low quality bases from 3' ends, or merges overlapping pairs into consensus                                                                                                       |
| [AfterQC](modules/afterqc.md)                                 | Automatic filtering, trimming, error removing, and quality control for FastQ data                                                                                                                                 |
| [Anglerfish](modules/anglerfish.md)                           | Quality controls Illumina libraries sequenced on Oxford Nanopore flowcells                                                                                                                                        |
| [Bakta](modules/bakta.md)                                     | Rapid & standardized annotation of bacterial genomes, MAGs & plasmids                                                                                                                                             |
| [Bamdst](modules/bamdst.md)                                   | Lightweight tool to stat the depth coverage of target regions of BAM file(s)                                                                                                                                      |
| [Bamtools](modules/bamtools.md)                               | Provides both a programmer's API and an end-user's toolkit for handling BAM files                                                                                                                                 |
| [BBDuk](modules/bbduk.md)                                     | Common data-quality-related trimming, filtering, and masking operations with a kmer based approach                                                                                                                |
| [BBTools](modules/bbmap.md)                                   | Pre-processing, assembly, alignment, and statistics tools for DNA/RNA sequencing reads                                                                                                                            |
| [Bcftools](modules/bcftools.md)                               | Utilities for variant calling and manipulating VCFs and BCFs                                                                                                                                                      |
| [bcl2fastq](modules/bcl2fastq.md)                             | Demultiplexes data and converts BCL files to FASTQ file formats for downstream analysis                                                                                                                           |
| [BCL Convert](modules/bclconvert.md)                          | Demultiplexes data and converts BCL files to FASTQ file formats for downstream analysis                                                                                                                           |
| [biobambam2](modules/biobambam2.md)                           | Tools for early stage alignment file processing                                                                                                                                                                   |
| [BioBloom Tools](modules/biobloomtools.md)                    | Assigns reads to different references using bloom filters. This is faster than alignment and can be used for contamination detection                                                                              |
| [BISCUIT](modules/biscuit.md)                                 | Maps bisulfite converted DNA sequence reads and determines cytosine methylation states                                                                                                                            |
| [Bismark](modules/bismark.md)                                 | Maps bisulfite converted sequence reads and determine cytosine methylation states                                                                                                                                 |
| [Bowtie 1](modules/bowtie1.md)                                | Ultrafast, memory-efficient short read aligner                                                                                                                                                                    |
| [Bowtie 2 / HiSAT2](modules/bowtie2.md)                       | Results from both Bowtie 2 and HISAT2, tools for aligning reads against a reference genome                                                                                                                        |
| [BUSCO](modules/busco.md)                                     | Assesses genome assembly and annotation completeness                                                                                                                                                              |
| [Bustools](modules/bustools.md)                               | Tools for BUS files - a file format for single-cell RNA-seq data designed to facilitate the development of modular workflows for data processing                                                                  |
| [CCS](modules/ccs.md)                                         | PacBio tool that generates highly accurate single-molecule consensus reads (HiFi Reads)                                                                                                                           |
| [Cell Ranger](modules/cellranger.md)                          | Analyzes single cell expression or VDJ data produced by 10X Genomics                                                                                                                                              |
| [CheckQC](modules/checkqc.md)                                 | Checks a set of quality criteria against an Illumina runfolder                                                                                                                                                    |
| [ClipAndMerge](modules/clipandmerge.md)                       | Adapter clipping and read merging for ancient DNA data                                                                                                                                                            |
| [Cluster Flow](modules/clusterflow.md)                        | Simple and flexible bioinformatics pipeline tool                                                                                                                                                                  |
| [Conpair](modules/conpair.md)                                 | Estimates concordance and contamination for tumorâ€“normal pairs                                                                                                                                                    |
| [Cutadapt](modules/cutadapt.md)                               | Finds and removes adapter sequences, primers, poly-A tails, and other types of unwanted sequences                                                                                                                 |
| [DamageProfiler](modules/damageprofiler.md)                   | DNA damage pattern retrieval for ancient DNA analysis                                                                                                                                                             |
| [DeDup](modules/dedup.md)                                     | Improved Duplicate Removal for merged/collapsed reads in ancient DNA analysis                                                                                                                                     |
| [deepTools](modules/deeptools.md)                             | Tools to process and analyze deep sequencing data                                                                                                                                                                 |
| [DIAMOND](modules/diamond.md)                                 | Sequence aligner for protein and translated DNA searches, a drop-in replacement for the NCBI BLAST                                                                                                                |
| [Disambiguate](modules/disambiguate.md)                       | Disambiguate reads aligned to two different species (e.g. human and mouse)                                                                                                                                        |
| [DRAGEN](modules/dragen.md)                                   | Illumina Bio-IT Platform that uses FPGA for secondary analysis of sequencing data                                                                                                                                 |
| [DRAGEN-FastQC](modules/dragen_fastqc.md)                     | Illumina Bio-IT Platform that uses FPGA for secondary analysis of sequencing data                                                                                                                                 |
| [eigenstratdatabasetools](modules/eigenstratdatabasetools.md) | Tools to compare and manipulate the contents of EingenStrat databases, and to calculate SNP coverage statistics in such databases                                                                                 |
| [fastp](modules/fastp.md)                                     | All-in-one FASTQ preprocessor (QC, adapters, trimming, filtering, splitting...)                                                                                                                                   |
| [FastQ Screen](modules/fastq_screen.md)                       | Screens a library of sequences in FastQ format against a set of sequence databases to see if the composition of the library matches with what you expect                                                          |
| [FastQC](modules/fastqc.md)                                   | Quality control tool for high throughput sequencing data                                                                                                                                                          |
| [featureCounts](modules/featurecounts.md)                     | Counts mapped reads for genomic features such as genes, exons, promoter, gene bodies, genomic bins and chromosomal locations                                                                                      |
| [fgbio](modules/fgbio.md)                                     | Processing and evaluating data containing UMIs                                                                                                                                                                    |
| [Filtlong](modules/filtlong.md)                               | Filters long reads by quality                                                                                                                                                                                     |
| [FLASh](modules/flash.md)                                     | Merges paired-end reads from next-generation sequencing experiments                                                                                                                                               |
| [Flexbar](modules/flexbar.md)                                 | Barcode and adapter removal tool                                                                                                                                                                                  |
| [Freyja](modules/freyja.md)                                   | Recovers relative lineage abundances from mixed SARS-CoV-2 samples                                                                                                                                                |
| [Ganon](modules/ganon.md)                                     | Metagenomics classification: quickly assigns sequence fragments to their closest reference among thousands of references via Interleaved Bloom Filters of k-mer/minimizers                                        |
| [GATK](modules/gatk.md)                                       | Wide variety of tools with a primary focus on variant discovery and genotyping                                                                                                                                    |
| [GffCompare](modules/gffcompare.md)                           | Tool to compare, merge and annotate one or more GFF files with a reference annotation in GFF format                                                                                                               |
| [GLIMPSE](modules/glimpse.md)                                 | Low-coverage whole genome sequencing imputation                                                                                                                                                                   |
| [goleft indexcov](modules/goleft_indexcov.md)                 | Quickly estimate coverage from a whole-genome bam index, providing 16KB resolution                                                                                                                                |
| [GoPeaks](modules/gopeaks.md)                                 | Calls peaks in CUT&TAG/CUT&RUN datasets                                                                                                                                                                           |
| [hap.py](modules/happy.md)                                    | Benchmarks variant calls against gold standard truth datasets                                                                                                                                                     |
| [HiCExplorer](modules/hicexplorer.md)                         | Hi-C analysis from processing to visualization                                                                                                                                                                    |
| [HiC-Pro](modules/hicpro.md)                                  | Pipeline for Hi-C data processing                                                                                                                                                                                 |
| [HiCUP](modules/hicup.md)                                     | Mapping and quality control on Hi-C data                                                                                                                                                                          |
| [HiFiasm](modules/hifiasm.md)                                 | Haplotype-resolved assembler for accurate Hifi reads                                                                                                                                                              |
| [HISAT2](modules/hisat2.md)                                   | Maps DNA or RNA reads against a genome or a population of genomes                                                                                                                                                 |
| [HOMER](modules/homer.md)                                     | Motif discovery and next-gen sequencing analysis                                                                                                                                                                  |
| [HOPS](modules/hops.md)                                       | Ancient DNA characteristics screening tool of output from the metagenomic aligner MALT                                                                                                                            |
| [Hostile](modules/hostile.md)                                 | Removes host sequences from short and long read (meta)genomes, from paired or unpaired fastq[.gz]                                                                                                                 |
| [HTSeq Count](modules/htseq.md)                               | Part of the HTSeq package: counts reads covering specified genomic features                                                                                                                                       |
| [HUMID](modules/humid.md)                                     | Reference-free tool to quickly remove duplicates from FastQ files, with or without UMIs                                                                                                                           |
| [Illumina InterOp Statistics](modules/interop.md)             | Reading and writing InterOp metric files                                                                                                                                                                          |
| [Iso-Seq](modules/isoseq.md)                                  | Identifies transcripts in PacBio single-molecule sequencing data (HiFi reads)                                                                                                                                     |
| [iVar](modules/ivar.md)                                       | Functions for viral amplicon-based sequencing                                                                                                                                                                     |
| [JCVI Genome Annotation](modules/jcvi.md)                     | Computes statistics on genome annotation                                                                                                                                                                          |
| [Jellyfish](modules/jellyfish.md)                             | Counting k-mers in DNA                                                                                                                                                                                            |
| [Kaiju](modules/kaiju.md)                                     | Taxonomic classification for metagenomics                                                                                                                                                                         |
| [Kallisto](modules/kallisto.md)                               | Quantifies abundances of transcripts (or more generally, of target sequences) from RNA-Seq data                                                                                                                   |
| [K-mer Analysis Toolkit](modules/kat.md)                      | Analyses sequencing data via its k-mer spectra                                                                                                                                                                    |
| [Kraken](modules/kraken.md)                                   | Taxonomic classification using exact k-mer matches to find the lowest common ancestor (LCA) of a given sequence                                                                                                   |
| [leeHom](modules/leehom.md)                                   | Bayesian reconstruction of ancient DNA                                                                                                                                                                            |
| [Librarian](modules/librarian.md)                             | Predicts the sequencing library type from the base composition of a FastQ file                                                                                                                                    |
| [Lima](modules/lima.md)                                       | Demultiplex PacBio single-molecule sequencing reads                                                                                                                                                               |
| [Long Ranger](modules/longranger.md)                          | Sample demultiplexing, barcode processing, alignment, quality control, variant calling, phasing, and structural variant calling                                                                                   |
| [MACS2](modules/macs2.md)                                     | Identifies transcription factor binding sites in ChIP-seq data                                                                                                                                                    |
| [MALT](modules/malt.md)                                       | Aligns of metagenomic reads to a database of reference sequences (such as NR, GenBank or Silva) and outputs a MEGAN RMA file                                                                                      |
| [mapDamage](modules/mapdamage.md)                             | Tracks and quantifies damage patterns in ancient DNA sequences                                                                                                                                                    |
| [MEGAHIT](modules/megahit.md)                                 | NGS read assembler                                                                                                                                                                                                |
| [MetaPhlAn](modules/metaphlan.md)                             | Profiles the composition of microbial communities from metagenomic shotgun sequencing data                                                                                                                        |
| [methylQA](modules/methylqa.md)                               | Methylation sequencing data quality assessment tool                                                                                                                                                               |
| [MinIONQC](modules/minionqc.md)                               | Quality control for ONT (Oxford Nanopore) long reads                                                                                                                                                              |
| [mirtop](modules/mirtop.md)                                   | Annotates miRNAs and isomiRs and compute general statistics in mirGFF3 format                                                                                                                                     |
| [miRTrace](modules/mirtrace.md)                               | Quality control for small RNA sequencing data                                                                                                                                                                     |
| [Mosdepth](modules/mosdepth.md)                               | Fast BAM/CRAM depth calculation for WGS, exome, or targeted sequencing                                                                                                                                            |
| [Motus](modules/motus.md)                                     | Microbial profiling through marker gene (MG)-based operational taxonomic units (mOTUs)                                                                                                                            |
| [mtnucratio](modules/mtnucratio.md)                           | Computes mitochondrial to nuclear genome ratios in NGS datasets                                                                                                                                                   |
| [MultiVCFAnalyzer](modules/multivcfanalyzer.md)               | Reads multiple VCF files into combined genotype calls, produces summary statistics and downstream formats                                                                                                         |
| [nanoq](modules/nanoq.md)                                     | Reports read quality and length from nanopore sequencing data                                                                                                                                                     |
| [NanoStat](modules/nanostat.md)                               | Reports various statistics for long read dataset in FASTQ, BAM, or albacore sequencing summary format (supports NanoPack; NanoPlot, NanoComp)                                                                     |
| [Nextclade](modules/nextclade.md)                             | Viral genome alignment, clade assignment, mutation calling, and quality checks                                                                                                                                    |
| [ngs-bits](modules/ngsbits.md)                                | Calculating statistics from FASTQ, BAM, and VCF                                                                                                                                                                   |
| [ngsderive](modules/ngsderive.md)                             | Forensic tool for by backwards computing library information in sequencing data                                                                                                                                   |
| [Nonpareil](modules/nonpareil.md)                             | Estimates metagenomic coverage and sequence diversity                                                                                                                                                             |
| [ODGI](modules/odgi.md)                                       | Analysis and manipulation of pangenome graphs structured in the variation graph model                                                                                                                             |
| [OptiType](modules/optitype.md)                               | Precision HLA typing from next-generation sequencing data                                                                                                                                                         |
| [pairtools](modules/pairtools.md)                             | Toolkit for Chromatin Conformation Capture experiments. Handles short-reads paired reference alignments, extracts 3C-specific information, and perform common tasks such as sorting, filtering, and deduplication |
| [Pangolin](modules/pangolin.md)                               | Uses variant calls to assign SARS-CoV-2 genome sequences to global lineages                                                                                                                                       |
| [pbmarkdup](modules/pbmarkdup.md)                             | Takes one or multiple sequencing chips of an amplified libray as HiFi reads and marks or removes duplicates                                                                                                       |
| [Peddy](modules/peddy.md)                                     | Compares familial-relationships and sexes as reported in a PED file with those inferred from a VCF                                                                                                                |
| [phantompeakqualtools](modules/phantompeakqualtools.md)       | Computes informative enrichment and quality measures for ChIP-seq/DNase-seq/FAIRE-seq/MNase-seq data                                                                                                              |
| [Picard](modules/picard.md)                                   | Tools for manipulating high-throughput sequencing data                                                                                                                                                            |
| [Porechop](modules/porechop.md)                               | Finds and removes adapters from Oxford Nanopore reads                                                                                                                                                             |
| [Preseq](modules/preseq.md)                                   | Estimates library complexity, showing how many additional unique reads are sequenced for increasing total read count                                                                                              |
| [PRINSEQ++](modules/prinseqplusplus.md)                       | C++ implementation of the prinseq-lite.pl program. Filters, reformats, and trims genomic and metagenomic reads                                                                                                    |
| [Prokka](modules/prokka.md)                                   | Rapid annotation of prokaryotic genomes                                                                                                                                                                           |
| [PURPLE](modules/purple.md)                                   | A purity, ploidy and copy number estimator for whole genome tumor data                                                                                                                                            |
| [Pychopper](modules/pychopper.md)                             | Identifies, orients, trims and rescues full length Nanopore cDNA reads. Can also rescue fused reads                                                                                                               |
| [pycoQC](modules/pycoqc.md)                                   | Computes metrics and generates interactive QC plots for Oxford Nanopore technologies sequencing data                                                                                                              |
| [qc3C](modules/qc3C.md)                                       | Reference-free and BAM based quality control for Hi-C data                                                                                                                                                        |
| [QoRTs](modules/qorts.md)                                     | Toolkit for analysis, QC, and data management of RNA-Seq datasets                                                                                                                                                 |
| [QualiMap](modules/qualimap.md)                               | Quality control of alignment data and its derivatives like feature counts                                                                                                                                         |
| [QUAST](modules/quast.md)                                     | Quality assessment tool for genome assemblies                                                                                                                                                                     |
| [RNA-SeQC](modules/rna_seqc.md)                               | RNA-Seq metrics for quality control and process optimization                                                                                                                                                      |
| [Rockhopper](modules/rockhopper.md)                           | Bacterial RNA-seq analysis: align reads to coding sequences, rRNAs, tRNAs, and miscellaneous RNAs                                                                                                                 |
| [RSEM](modules/rsem.md)                                       | Estimates gene and isoform expression levels from RNA-Seq data                                                                                                                                                    |
| [RSeQC](modules/rseqc.md)                                     | Evaluates high throughput RNA-seq data                                                                                                                                                                            |
| [Salmon](modules/salmon.md)                                   | Quantifies expression of transcripts using RNA-seq data                                                                                                                                                           |
| [Sambamba](modules/sambamba.md)                               | Toolkit for interacting with BAM/CRAM files                                                                                                                                                                       |
| [Samblaster](modules/samblaster.md)                           | Marks duplicates and extracts discordant and split reads from sam files                                                                                                                                           |
| [Samtools](modules/samtools.md)                               | Toolkit for interacting with BAM/CRAM files                                                                                                                                                                       |
| [Sargasso](modules/sargasso.md)                               | Separates mixed-species RNA-seq reads according to their species of origin                                                                                                                                        |
| [Seqera Platform CLI](modules/seqera_cli.md)                  | Reports statistics generated by the Seqera Platform CLI                                                                                                                                                           |
| [Sequali](modules/sequali.md)                                 | Sequencing quality control for both long-read and short-read data                                                                                                                                                 |
| [SeqWho](modules/seqwho.md)                                   | Determines FASTQ(A) sequencing file source protocol and the species of origin, to check that the composition of the library is expected                                                                           |
| [SeqyClean](modules/seqyclean.md)                             | Filters adapters, vectors, and contaminants while quality trimming                                                                                                                                                |
| [SexDetErrmine](modules/sexdeterrmine.md)                     | Calculates relative coverage of X and Y chromosomes and their associated error bars from the depth of coverage at specified SNPs                                                                                  |
| [Sickle](modules/sickle.md)                                   | A windowed adaptive trimming tool for FASTQ files using quality                                                                                                                                                   |
| [Skewer](modules/skewer.md)                                   | Adapter trimming tool for NGS paired-end sequences                                                                                                                                                                |
| [Slamdunk](modules/slamdunk.md)                               | Tool to analyze SLAM-Seq data                                                                                                                                                                                     |
| [Snippy](modules/snippy.md)                                   | Rapid haploid variant calling and core genome alignment                                                                                                                                                           |
| [SnpEff](modules/snpeff.md)                                   | Annotates and predicts the effects of variants on genes (such as amino acid changes)                                                                                                                              |
| [SNPsplit](modules/snpsplit.md)                               | Allele-specific alignment sorter. Determines allelic origin of reads that cover known SNP positions                                                                                                               |
| [Somalier](modules/somalier.md)                               | Genotype to pedigree correspondence checks from sketches derived from BAM/CRAM or VCF                                                                                                                             |
| [SortMeRNA](modules/sortmerna.md)                             | Program for filtering, mapping and OTU-picking NGS reads in metatranscriptomic and metagenomic data                                                                                                               |
| [Sourmash](modules/sourmash.md)                               | Quickly searches, compares, and analyzes genomic and metagenomic data sets                                                                                                                                        |
| [Space Ranger](modules/spaceranger.md)                        | Tool to analyze 10x Genomics spatial transcriptomics data                                                                                                                                                         |
| [Stacks](modules/stacks.md)                                   | Analyzes restriction enzyme-based data (e.g. RAD-seq)                                                                                                                                                             |
| [STAR](modules/star.md)                                       | Universal RNA-seq aligner                                                                                                                                                                                         |
| [Supernova](modules/supernova.md)                             | De novo genome assembler of 10X Genomics linked-reads                                                                                                                                                             |
| [THetA2](modules/theta2.md)                                   | Estimates tumour purity and clonal / subclonal copy number                                                                                                                                                        |
| [Tophat](modules/tophat.md)                                   | Splice junction RNA-Seq reads mapper for mammalian-sized genomes                                                                                                                                                  |
| [Trimmomatic](modules/trimmomatic.md)                         | Read trimming tool for Illumina NGS data                                                                                                                                                                          |
| [Truvari](modules/truvari.md)                                 | Benchmarking, merging, and annotating structural variants                                                                                                                                                         |
| [UMICollapse](modules/umicollapse.md)                         | Algorithms for efficiently collapsing reads with Unique Molecular Identifiers                                                                                                                                     |
| [UMI-tools](modules/umitools.md)                              | Tools for dealing with Unique Molecular Identifiers (UMIs)/(RMTs) and scRNA-Seq barcodes                                                                                                                          |
| [VarScan2](modules/varscan2.md)                               | Variant detection in massively parallel sequencing data                                                                                                                                                           |
| [VCFTools](modules/vcftools.md)                               | Program to analyse and reporting on VCF files                                                                                                                                                                     |
| [VEP](modules/vep.md)                                         | Determines the effect of variants on genes, transcripts and protein sequences, as well as regulatory regions                                                                                                      |
| [VerifyBAMID](modules/verifybamid.md)                         | Detects sample contamination and/or sample swaps                                                                                                                                                                  |
| [VG](modules/vg.md)                                           | Toolkit to manipulate and analyze graphical genomes, including read alignment                                                                                                                                     |
| [WhatsHap](modules/whatshap.md)                               | Phasing genomic variants using DNA reads (aka read-based phasing, or haplotype assembly)                                                                                                                          |
| [Xengsort](modules/xengsort.md)                               | Fast xenograft read sorter based on space-efficient k-mer hashing                                                                                                                                                 |
| [Xenome](modules/xenome.md)                                   | Classifies reads from xenograft sources                                                                                                                                                                           |
