---
title: Supported Tools
description: Tools supported by MultiQC
displayed_sidebar: multiqcSidebar
---

<!--
~~~~~ DO NOT EDIT ~~~~~
This file is autogenerated. Do not edit the markdown, it will be overwritten.
~~~~~~~~~~~~~~~~~~~~~~~
-->

MultiQC currently has modules to support 167 bioinformatics tools, listed below.

Click the tool name to go to the MultiQC documentation for that tool.

:::tip[Missing something?]
If you would like another tool to to be supported, please [open an issue](https://github.com/MultiQC/MultiQC/issues/new?labels=module%3A+new&template=module-request.yml).
:::

import MultiqcModules from "@site/src/components/MultiqcModules";

<MultiqcModules
  modules={[
    {
      id: "modules/adapterremoval",
      data: {
        name: "Adapter Removal",
        summary:
          "<p>Removes adapter sequences, trims low quality bases from 3' ends, or merges overlapping pairs into consensus.</p>",
      },
    },
    {
      id: "modules/afterqc",
      data: {
        name: "AfterQC",
        summary: "<p>Automatic filtering, trimming, error removing, and quality control for FastQ data.</p>",
      },
    },
    {
      id: "modules/anglerfish",
      data: {
        name: "Anglerfish",
        summary: "<p>Quality controls Illumina libraries sequenced on Oxford Nanopore flowcells.</p>",
      },
    },
    {
      id: "modules/ataqv",
      data: { name: "ATAQV", summary: "<p>Toolkit for quality control and visualization of ATAC-seq data.</p>" },
    },
    {
      id: "modules/bakta",
      data: {
        name: "Bakta",
        summary: "<p>Rapid &amp; standardized annotation of bacterial genomes, MAGs &amp; plasmids.</p>",
      },
    },
    {
      id: "modules/bamdst",
      data: {
        name: "Bamdst",
        summary: "<p>Lightweight tool to stat the depth coverage of target regions of BAM file(s).</p>",
      },
    },
    {
      id: "modules/bamtools",
      data: {
        name: "Bamtools",
        summary: "<p>Provides both a programmer's API and an end-user's toolkit for handling BAM files.</p>",
      },
    },
    {
      id: "modules/bases2fastq",
      data: {
        name: "Bases2Fastq",
        summary: "<p>Demultiplexes and converts Element AVITI base calls into FASTQ files.</p>",
      },
    },
    {
      id: "modules/bbduk",
      data: {
        name: "BBDuk",
        summary:
          "<p>Common data-quality-related trimming, filtering, and masking operations with a kmer based approach.</p>",
      },
    },
    {
      id: "modules/bbmap",
      data: {
        name: "BBTools",
        summary: "<p>Pre-processing, assembly, alignment, and statistics tools for DNA/RNA sequencing reads.</p>",
      },
    },
    {
      id: "modules/bcftools",
      data: { name: "Bcftools", summary: "<p>Utilities for variant calling and manipulating VCFs and BCFs.</p>" },
    },
    {
      id: "modules/bcl2fastq",
      data: {
        name: "bcl2fastq",
        summary: "<p>Demultiplexes data and converts BCL files to FASTQ file formats for downstream analysis.</p>",
      },
    },
    {
      id: "modules/bclconvert",
      data: {
        name: "BCL Convert",
        summary: "<p>Demultiplexes data and converts BCL files to FASTQ file formats for downstream analysis.</p>",
      },
    },
    {
      id: "modules/biobambam2",
      data: { name: "biobambam2", summary: "<p>Tools for early stage alignment file processing.</p>" },
    },
    {
      id: "modules/biobloomtools",
      data: {
        name: "BioBloom Tools",
        summary:
          "<p>Assigns reads to different references using bloom filters. This is faster than alignment and can be used for contamination detection.</p>",
      },
    },
    {
      id: "modules/biscuit",
      data: {
        name: "BISCUIT",
        summary: "<p>Maps bisulfite converted DNA sequence reads and determines cytosine methylation states.</p>",
      },
    },
    {
      id: "modules/bismark",
      data: {
        name: "Bismark",
        summary: "<p>Maps bisulfite converted sequence reads and determine cytosine methylation states.</p>",
      },
    },
    {
      id: "modules/bowtie1",
      data: { name: "Bowtie 1", summary: "<p>Ultrafast, memory-efficient short read aligner.</p>" },
    },
    {
      id: "modules/bowtie2",
      data: {
        name: "Bowtie 2 / HiSAT2",
        summary: "<p>Results from both Bowtie 2 and HISAT2, tools for aligning reads against a reference genome.</p>",
      },
    },
    {
      id: "modules/busco",
      data: { name: "BUSCO", summary: "<p>Assesses genome assembly and annotation completeness.</p>" },
    },
    {
      id: "modules/bustools",
      data: {
        name: "Bustools",
        summary:
          "<p>Tools for BUS files - a file format for single-cell RNA-seq data designed to facilitate the development of modular workflows for data processing.</p>",
      },
    },
    {
      id: "modules/ccs",
      data: {
        name: "CCS",
        summary: "<p>PacBio tool that generates highly accurate single-molecule consensus reads (HiFi Reads).</p>",
      },
    },
    {
      id: "modules/cellranger",
      data: {
        name: "Cell Ranger",
        summary: "<p>Analyzes single cell expression or VDJ data produced by 10X Genomics.</p>",
      },
    },
    {
      id: "modules/cellranger_arc",
      data: {
        name: "Cell Ranger ARC",
        summary: "<p>Analyzes single-cell multiome ATAC and gene expression data produced by 10X Genomics.</p>",
      },
    },
    {
      id: "modules/cells2stats",
      data: {
        name: "cells2stats",
        summary: "<p>Generate output files and statistics from Element Biosciences Teton cytoprofiling assays.</p>",
      },
    },
    {
      id: "modules/checkm",
      data: {
        name: "CheckM",
        summary:
          "<p>Estimates genome completeness and contamination based on the presence or absence of marker genes.</p>",
      },
    },
    {
      id: "modules/checkm2",
      data: { name: "CheckM2", summary: "<p>Assesses microbial genome quality using machine learning.</p>" },
    },
    {
      id: "modules/checkqc",
      data: { name: "CheckQC", summary: "<p>Checks a set of quality criteria against an Illumina runfolder.</p>" },
    },
    {
      id: "modules/clipandmerge",
      data: { name: "ClipAndMerge", summary: "<p>Adapter clipping and read merging for ancient DNA data.</p>" },
    },
    {
      id: "modules/clusterflow",
      data: { name: "Cluster Flow", summary: "<p>Simple and flexible bioinformatics pipeline tool.</p>" },
    },
    {
      id: "modules/conpair",
      data: { name: "Conpair", summary: "<p>Estimates concordance and contamination for tumor\u2013normal pairs.</p>" },
    },
    {
      id: "modules/cutadapt",
      data: {
        name: "Cutadapt",
        summary:
          "<p>Finds and removes adapter sequences, primers, poly-A tails, and other types of unwanted sequences.</p>",
      },
    },
    {
      id: "modules/damageprofiler",
      data: { name: "DamageProfiler", summary: "<p>DNA damage pattern retrieval for ancient DNA analysis.</p>" },
    },
    {
      id: "modules/dedup",
      data: {
        name: "DeDup",
        summary: "<p>Improved Duplicate Removal for merged/collapsed reads in ancient DNA analysis.</p>",
      },
    },
    {
      id: "modules/deeptools",
      data: { name: "deepTools", summary: "<p>Tools to process and analyze deep sequencing data.</p>" },
    },
    {
      id: "modules/diamond",
      data: {
        name: "DIAMOND",
        summary:
          "<p>Sequence aligner for protein and translated DNA searches, a drop-in replacement for the NCBI BLAST.</p>",
      },
    },
    {
      id: "modules/disambiguate",
      data: {
        name: "Disambiguate",
        summary: "<p>Disambiguate reads aligned to two different species (e.g. human and mouse).</p>",
      },
    },
    {
      id: "modules/dragen",
      data: {
        name: "DRAGEN",
        summary: "<p>Illumina Bio-IT Platform that uses FPGA for secondary analysis of sequencing data.</p>",
      },
    },
    {
      id: "modules/dragen_fastqc",
      data: {
        name: "DRAGEN-FastQC",
        summary: "<p>Illumina Bio-IT Platform that uses FPGA for secondary analysis of sequencing data.</p>",
      },
    },
    {
      id: "modules/eigenstratdatabasetools",
      data: {
        name: "eigenstratdatabasetools",
        summary:
          "<p>Tools to compare and manipulate the contents of EingenStrat databases, and to calculate SNP coverage statistics in such databases.</p>",
      },
    },
    {
      id: "modules/fastp",
      data: {
        name: "fastp",
        summary: "<p>All-in-one FASTQ preprocessor (QC, adapters, trimming, filtering, splitting...).</p>",
      },
    },
    {
      id: "modules/fastq_screen",
      data: {
        name: "FastQ Screen",
        summary:
          "<p>Screens a library of sequences in FastQ format against a set of sequence databases to see if the composition of the library matches with what you expect.</p>",
      },
    },
    {
      id: "modules/fastqc",
      data: { name: "FastQC", summary: "<p>Quality control tool for high throughput sequencing data.</p>" },
    },
    {
      id: "modules/featurecounts",
      data: {
        name: "featureCounts",
        summary:
          "<p>Counts mapped reads for genomic features such as genes, exons, promoter, gene bodies, genomic bins and chromosomal locations.</p>",
      },
    },
    { id: "modules/fgbio", data: { name: "fgbio", summary: "<p>Processing and evaluating data containing UMIs.</p>" } },
    { id: "modules/filtlong", data: { name: "Filtlong", summary: "<p>Filters long reads by quality.</p>" } },
    {
      id: "modules/flash",
      data: { name: "FLASh", summary: "<p>Merges paired-end reads from next-generation sequencing experiments.</p>" },
    },
    { id: "modules/flexbar", data: { name: "Flexbar", summary: "<p>Barcode and adapter removal tool.</p>" } },
    {
      id: "modules/freyja",
      data: { name: "Freyja", summary: "<p>Recovers relative lineage abundances from mixed SARS-CoV-2 samples.</p>" },
    },
    {
      id: "modules/ganon",
      data: {
        name: "Ganon",
        summary:
          "<p>Metagenomics classification: quickly assigns sequence fragments to their closest reference among thousands of references via Interleaved Bloom Filters of k-mer/minimizers.</p>",
      },
    },
    {
      id: "modules/gatk",
      data: {
        name: "GATK",
        summary: "<p>Wide variety of tools with a primary focus on variant discovery and genotyping.</p>",
      },
    },
    {
      id: "modules/gffcompare",
      data: {
        name: "GffCompare",
        summary:
          "<p>Tool to compare, merge and annotate one or more GFF files with a reference annotation in GFF format.</p>",
      },
    },
    {
      id: "modules/glimpse",
      data: { name: "GLIMPSE", summary: "<p>Low-coverage whole genome sequencing imputation.</p>" },
    },
    {
      id: "modules/goleft_indexcov",
      data: {
        name: "goleft indexcov",
        summary: "<p>Quickly estimate coverage from a whole-genome bam index, providing 16KB resolution.</p>",
      },
    },
    {
      id: "modules/gopeaks",
      data: { name: "GoPeaks", summary: "<p>Calls peaks in CUT&amp;TAG/CUT&amp;RUN datasets.</p>" },
    },
    {
      id: "modules/gtdbtk",
      data: {
        name: "GTDB-Tk",
        summary: "<p>Assigns objective taxonomic classifications to bacterial and archaeal genomes.</p>",
      },
    },
    {
      id: "modules/haplocheck",
      data: {
        name: "Haplocheck",
        summary:
          "<p>Detects in-sample contamination in mtDNA or WGS sequencing studies by analyzing the mitchondrial content.</p>",
      },
    },
    {
      id: "modules/happy",
      data: { name: "hap.py", summary: "<p>Benchmarks variant calls against gold standard truth datasets.</p>" },
    },
    {
      id: "modules/hicexplorer",
      data: { name: "HiCExplorer", summary: "<p>Hi-C analysis from processing to visualization.</p>" },
    },
    { id: "modules/hicpro", data: { name: "HiC-Pro", summary: "<p>Pipeline for Hi-C data processing.</p>" } },
    { id: "modules/hicup", data: { name: "HiCUP", summary: "<p>Mapping and quality control on Hi-C data.</p>" } },
    {
      id: "modules/hifiasm",
      data: { name: "HiFiasm", summary: "<p>Haplotype-resolved assembler for accurate Hifi reads.</p>" },
    },
    {
      id: "modules/hisat2",
      data: { name: "HISAT2", summary: "<p>Maps DNA or RNA reads against a genome or a population of genomes.</p>" },
    },
    {
      id: "modules/homer",
      data: { name: "HOMER", summary: "<p>Motif discovery and next-gen sequencing analysis.</p>" },
    },
    {
      id: "modules/hops",
      data: {
        name: "HOPS",
        summary: "<p>Ancient DNA characteristics screening tool of output from the metagenomic aligner MALT.</p>",
      },
    },
    {
      id: "modules/hostile",
      data: {
        name: "Hostile",
        summary:
          "<p>Removes host sequences from short and long read (meta)genomes, from paired or unpaired fastq[.gz].</p>",
      },
    },
    {
      id: "modules/htseq",
      data: {
        name: "HTSeq Count",
        summary: "<p>Part of the HTSeq package: counts reads covering specified genomic features.</p>",
      },
    },
    {
      id: "modules/humid",
      data: {
        name: "HUMID",
        summary: "<p>Reference-free tool to quickly remove duplicates from FastQ files, with or without UMIs.</p>",
      },
    },
    {
      id: "modules/interop",
      data: { name: "Illumina InterOp Statistics", summary: "<p>Reading and writing InterOp metric files.</p>" },
    },
    {
      id: "modules/isoseq",
      data: {
        name: "Iso-Seq",
        summary: "<p>Identifies transcripts in PacBio single-molecule sequencing data (HiFi reads).</p>",
      },
    },
    { id: "modules/ivar", data: { name: "iVar", summary: "<p>Functions for viral amplicon-based sequencing.</p>" } },
    {
      id: "modules/jcvi",
      data: { name: "JCVI Genome Annotation", summary: "<p>Computes statistics on genome annotation.</p>" },
    },
    { id: "modules/jellyfish", data: { name: "Jellyfish", summary: "<p>Counting k-mers in DNA.</p>" } },
    { id: "modules/kaiju", data: { name: "Kaiju", summary: "<p>Taxonomic classification for metagenomics.</p>" } },
    {
      id: "modules/kallisto",
      data: {
        name: "Kallisto",
        summary:
          "<p>Quantifies abundances of transcripts (or more generally, of target sequences) from RNA-Seq data.</p>",
      },
    },
    {
      id: "modules/kat",
      data: { name: "K-mer Analysis Toolkit", summary: "<p>Analyses sequencing data via its k-mer spectra.</p>" },
    },
    {
      id: "modules/kraken",
      data: {
        name: "Kraken",
        summary:
          "<p>Taxonomic classification using exact k-mer matches to find the lowest common ancestor (LCA) of a given sequence.</p>",
      },
    },
    { id: "modules/leehom", data: { name: "leeHom", summary: "<p>Bayesian reconstruction of ancient DNA.</p>" } },
    {
      id: "modules/librarian",
      data: {
        name: "Librarian",
        summary: "<p>Predicts the sequencing library type from the base composition of a FastQ file.</p>",
      },
    },
    {
      id: "modules/lima",
      data: { name: "Lima", summary: "<p>Demultiplex PacBio single-molecule sequencing reads.</p>" },
    },
    {
      id: "modules/longranger",
      data: {
        name: "Long Ranger",
        summary:
          "<p>Sample demultiplexing, barcode processing, alignment, quality control, variant calling, phasing, and structural variant calling.</p>",
      },
    },
    {
      id: "modules/macs2",
      data: { name: "MACS2", summary: "<p>Identifies transcription factor binding sites in ChIP-seq data.</p>" },
    },
    {
      id: "modules/malt",
      data: {
        name: "MALT",
        summary:
          "<p>Aligns of metagenomic reads to a database of reference sequences (such as NR, GenBank or Silva) and outputs a MEGAN RMA file.</p>",
      },
    },
    {
      id: "modules/mapdamage",
      data: { name: "mapDamage", summary: "<p>Tracks and quantifies damage patterns in ancient DNA sequences.</p>" },
    },
    { id: "modules/megahit", data: { name: "MEGAHIT", summary: "<p>NGS read assembler.</p>" } },
    {
      id: "modules/metaphlan",
      data: {
        name: "MetaPhlAn",
        summary: "<p>Profiles the composition of microbial communities from metagenomic shotgun sequencing data.</p>",
      },
    },
    {
      id: "modules/methylqa",
      data: { name: "methylQA", summary: "<p>Methylation sequencing data quality assessment tool.</p>" },
    },
    {
      id: "modules/mgikit",
      data: { name: "mgikit", summary: "<p>Demultiplexes FASTQ files from an MGI sequencing instrument.</p>" },
    },
    {
      id: "modules/minionqc",
      data: { name: "MinIONQC", summary: "<p>Quality control for ONT (Oxford Nanopore) long reads.</p>" },
    },
    {
      id: "modules/mirtop",
      data: {
        name: "mirtop",
        summary: "<p>Annotates miRNAs and isomiRs and compute general statistics in mirGFF3 format.</p>",
      },
    },
    {
      id: "modules/mirtrace",
      data: { name: "miRTrace", summary: "<p>Quality control for small RNA sequencing data.</p>" },
    },
    {
      id: "modules/mosaicatcher",
      data: {
        name: "MosaiCatcher",
        summary:
          "<p>Counts strand-seq reads and classifies strand states of each chromosome in each cell using a Hidden Markov Model.</p>",
      },
    },
    {
      id: "modules/mosdepth",
      data: {
        name: "Mosdepth",
        summary: "<p>Fast BAM/CRAM depth calculation for WGS, exome, or targeted sequencing.</p>",
      },
    },
    {
      id: "modules/motus",
      data: {
        name: "Motus",
        summary: "<p>Microbial profiling through marker gene (MG)-based operational taxonomic units (mOTUs).</p>",
      },
    },
    {
      id: "modules/mtnucratio",
      data: { name: "mtnucratio", summary: "<p>Computes mitochondrial to nuclear genome ratios in NGS datasets.</p>" },
    },
    {
      id: "modules/multivcfanalyzer",
      data: {
        name: "MultiVCFAnalyzer",
        summary:
          "<p>Reads multiple VCF files into combined genotype calls, produces summary statistics and downstream formats.</p>",
      },
    },
    {
      id: "modules/nanoq",
      data: { name: "nanoq", summary: "<p>Reports read quality and length from nanopore sequencing data.</p>" },
    },
    {
      id: "modules/nanostat",
      data: {
        name: "NanoStat",
        summary:
          "<p>Reports various statistics for long read dataset in FASTQ, BAM, or albacore sequencing summary format (supports NanoPack; NanoPlot, NanoComp).</p>",
      },
    },
    {
      id: "modules/nextclade",
      data: {
        name: "Nextclade",
        summary: "<p>Viral genome alignment, clade assignment, mutation calling, and quality checks.</p>",
      },
    },
    {
      id: "modules/ngsbits",
      data: { name: "ngs-bits", summary: "<p>Calculating statistics from FASTQ, BAM, and VCF.</p>" },
    },
    {
      id: "modules/ngsderive",
      data: {
        name: "ngsderive",
        summary: "<p>Forensic tool for by backwards computing library information in sequencing data.</p>",
      },
    },
    {
      id: "modules/nonpareil",
      data: { name: "Nonpareil", summary: "<p>Estimates metagenomic coverage and sequence diversity.</p>" },
    },
    {
      id: "modules/odgi",
      data: {
        name: "ODGI",
        summary: "<p>Analysis and manipulation of pangenome graphs structured in the variation graph model.</p>",
      },
    },
    {
      id: "modules/optitype",
      data: { name: "OptiType", summary: "<p>Precision HLA typing from next-generation sequencing data.</p>" },
    },
    {
      id: "modules/pairtools",
      data: {
        name: "pairtools",
        summary:
          "<p>Toolkit for Chromatin Conformation Capture experiments. Handles short-reads paired reference alignments, extracts 3C-specific information, and perform common tasks such as sorting, filtering, and deduplication.</p>",
      },
    },
    {
      id: "modules/pangolin",
      data: {
        name: "Pangolin",
        summary: "<p>Uses variant calls to assign SARS-CoV-2 genome sequences to global lineages.</p>",
      },
    },
    {
      id: "modules/pbmarkdup",
      data: {
        name: "pbmarkdup",
        summary:
          "<p>Takes one or multiple sequencing chips of an amplified libray as HiFi reads and marks or removes duplicates.</p>",
      },
    },
    {
      id: "modules/peddy",
      data: {
        name: "Peddy",
        summary:
          "<p>Compares familial-relationships and sexes as reported in a PED file with those inferred from a VCF.</p>",
      },
    },
    {
      id: "modules/percolator",
      data: {
        name: "Percolator",
        summary: "<p>Semi-supervised learning for peptide identification from shotgun proteomics datasets.</p>",
      },
    },
    {
      id: "modules/phantompeakqualtools",
      data: {
        name: "phantompeakqualtools",
        summary:
          "<p>Computes informative enrichment and quality measures for ChIP-seq/DNase-seq/FAIRE-seq/MNase-seq data.</p>",
      },
    },
    {
      id: "modules/picard",
      data: { name: "Picard", summary: "<p>Tools for manipulating high-throughput sequencing data.</p>" },
    },
    {
      id: "modules/porechop",
      data: { name: "Porechop", summary: "<p>Finds and removes adapters from Oxford Nanopore reads.</p>" },
    },
    {
      id: "modules/preseq",
      data: {
        name: "Preseq",
        summary:
          "<p>Estimates library complexity, showing how many additional unique reads are sequenced for increasing total read count.</p>",
      },
    },
    {
      id: "modules/prinseqplusplus",
      data: {
        name: "PRINSEQ++",
        summary:
          "<p>C++ implementation of the prinseq-lite.pl program. Filters, reformats, and trims genomic and metagenomic reads.</p>",
      },
    },
    { id: "modules/prokka", data: { name: "Prokka", summary: "<p>Rapid annotation of prokaryotic genomes.</p>" } },
    {
      id: "modules/purple",
      data: {
        name: "PURPLE",
        summary: "<p>A purity, ploidy and copy number estimator for whole genome tumor data.</p>",
      },
    },
    {
      id: "modules/pychopper",
      data: {
        name: "Pychopper",
        summary:
          "<p>Identifies, orients, trims and rescues full length Nanopore cDNA reads. Can also rescue fused reads.</p>",
      },
    },
    {
      id: "modules/pycoqc",
      data: {
        name: "pycoQC",
        summary:
          "<p>Computes metrics and generates interactive QC plots for Oxford Nanopore technologies sequencing data.</p>",
      },
    },
    {
      id: "modules/qc3C",
      data: { name: "qc3C", summary: "<p>Reference-free and BAM based quality control for Hi-C data.</p>" },
    },
    {
      id: "modules/qorts",
      data: { name: "QoRTs", summary: "<p>Toolkit for analysis, QC, and data management of RNA-Seq datasets.</p>" },
    },
    {
      id: "modules/qualimap",
      data: {
        name: "QualiMap",
        summary: "<p>Quality control of alignment data and its derivatives like feature counts.</p>",
      },
    },
    { id: "modules/quast", data: { name: "QUAST", summary: "<p>Quality assessment tool for genome assemblies.</p>" } },
    {
      id: "modules/rna_seqc",
      data: { name: "RNA-SeQC", summary: "<p>RNA-Seq metrics for quality control and process optimization.</p>" },
    },
    {
      id: "modules/rockhopper",
      data: {
        name: "Rockhopper",
        summary:
          "<p>Bacterial RNA-seq analysis: align reads to coding sequences, rRNAs, tRNAs, and miscellaneous RNAs.</p>",
      },
    },
    {
      id: "modules/rsem",
      data: { name: "RSEM", summary: "<p>Estimates gene and isoform expression levels from RNA-Seq data.</p>" },
    },
    { id: "modules/rseqc", data: { name: "RSeQC", summary: "<p>Evaluates high throughput RNA-seq data.</p>" } },
    {
      id: "modules/salmon",
      data: { name: "Salmon", summary: "<p>Quantifies expression of transcripts using RNA-seq data.</p>" },
    },
    {
      id: "modules/sambamba",
      data: { name: "Sambamba", summary: "<p>Toolkit for interacting with BAM/CRAM files.</p>" },
    },
    {
      id: "modules/samblaster",
      data: {
        name: "Samblaster",
        summary: "<p>Marks duplicates and extracts discordant and split reads from sam files.</p>",
      },
    },
    {
      id: "modules/samtools",
      data: { name: "Samtools", summary: "<p>Toolkit for interacting with BAM/CRAM files.</p>" },
    },
    {
      id: "modules/sargasso",
      data: {
        name: "Sargasso",
        summary: "<p>Separates mixed-species RNA-seq reads according to their species of origin.</p>",
      },
    },
    {
      id: "modules/seqera_cli",
      data: { name: "Seqera Platform CLI", summary: "<p>Reports statistics generated by the Seqera Platform CLI.</p>" },
    },
    { id: "modules/seqfu", data: { name: "Seqfu", summary: "<p>Manipulate FASTA/FASTQ files.</p>" } },
    {
      id: "modules/sequali",
      data: { name: "Sequali", summary: "<p>Sequencing quality control for both long-read and short-read data.</p>" },
    },
    {
      id: "modules/seqwho",
      data: {
        name: "SeqWho",
        summary:
          "<p>Determines FASTQ(A) sequencing file source protocol and the species of origin, to check that the composition of the library is expected.</p>",
      },
    },
    {
      id: "modules/seqyclean",
      data: {
        name: "SeqyClean",
        summary: "<p>Filters adapters, vectors, and contaminants while quality trimming.</p>",
      },
    },
    {
      id: "modules/sexdeterrmine",
      data: {
        name: "SexDetErrmine",
        summary:
          "<p>Calculates relative coverage of X and Y chromosomes and their associated error bars from the depth of coverage at specified SNPs.</p>",
      },
    },
    {
      id: "modules/sickle",
      data: { name: "Sickle", summary: "<p>A windowed adaptive trimming tool for FASTQ files using quality.</p>" },
    },
    {
      id: "modules/skewer",
      data: { name: "Skewer", summary: "<p>Adapter trimming tool for NGS paired-end sequences.</p>" },
    },
    { id: "modules/slamdunk", data: { name: "Slamdunk", summary: "<p>Tool to analyze SLAM-Seq data.</p>" } },
    {
      id: "modules/snippy",
      data: { name: "Snippy", summary: "<p>Rapid haploid variant calling and core genome alignment.</p>" },
    },
    {
      id: "modules/snpeff",
      data: {
        name: "SnpEff",
        summary: "<p>Annotates and predicts the effects of variants on genes (such as amino acid changes).</p>",
      },
    },
    {
      id: "modules/snpsplit",
      data: {
        name: "SNPsplit",
        summary:
          "<p>Allele-specific alignment sorter. Determines allelic origin of reads that cover known SNP positions.</p>",
      },
    },
    {
      id: "modules/somalier",
      data: {
        name: "Somalier",
        summary: "<p>Genotype to pedigree correspondence checks from sketches derived from BAM/CRAM or VCF.</p>",
      },
    },
    {
      id: "modules/sortmerna",
      data: {
        name: "SortMeRNA",
        summary:
          "<p>Program for filtering, mapping and OTU-picking NGS reads in metatranscriptomic and metagenomic data.</p>",
      },
    },
    {
      id: "modules/sourmash",
      data: {
        name: "Sourmash",
        summary: "<p>Quickly searches, compares, and analyzes genomic and metagenomic data sets.</p>",
      },
    },
    {
      id: "modules/spaceranger",
      data: { name: "Space Ranger", summary: "<p>Tool to analyze 10x Genomics spatial transcriptomics data.</p>" },
    },
    {
      id: "modules/stacks",
      data: { name: "Stacks", summary: "<p>Analyzes restriction enzyme-based data (e.g. RAD-seq).</p>" },
    },
    { id: "modules/star", data: { name: "STAR", summary: "<p>Universal RNA-seq aligner.</p>" } },
    {
      id: "modules/supernova",
      data: { name: "Supernova", summary: "<p>De novo genome assembler of 10X Genomics linked-reads.</p>" },
    },
    {
      id: "modules/telseq",
      data: { name: "telseq", summary: "<p>Estimates telomere length from whole genome sequencing data (BAMs).</p>" },
    },
    {
      id: "modules/theta2",
      data: { name: "THetA2", summary: "<p>Estimates tumour purity and clonal / subclonal copy number.</p>" },
    },
    {
      id: "modules/tophat",
      data: { name: "Tophat", summary: "<p>Splice junction RNA-Seq reads mapper for mammalian-sized genomes.</p>" },
    },
    {
      id: "modules/trimmomatic",
      data: { name: "Trimmomatic", summary: "<p>Read trimming tool for Illumina NGS data.</p>" },
    },
    {
      id: "modules/truvari",
      data: { name: "Truvari", summary: "<p>Benchmarking, merging, and annotating structural variants.</p>" },
    },
    {
      id: "modules/umicollapse",
      data: {
        name: "UMICollapse",
        summary: "<p>Algorithms for efficiently collapsing reads with Unique Molecular Identifiers.</p>",
      },
    },
    {
      id: "modules/umitools",
      data: {
        name: "UMI-tools",
        summary: "<p>Tools for dealing with Unique Molecular Identifiers (UMIs)/(RMTs) and scRNA-Seq barcodes.</p>",
      },
    },
    {
      id: "modules/varscan2",
      data: { name: "VarScan2", summary: "<p>Variant detection in massively parallel sequencing data.</p>" },
    },
    {
      id: "modules/vcftools",
      data: { name: "VCFTools", summary: "<p>Program to analyse and reporting on VCF files.</p>" },
    },
    {
      id: "modules/vep",
      data: {
        name: "VEP",
        summary:
          "<p>Determines the effect of variants on genes, transcripts and protein sequences, as well as regulatory regions.</p>",
      },
    },
    {
      id: "modules/verifybamid",
      data: { name: "VerifyBAMID", summary: "<p>Detects sample contamination and/or sample swaps.</p>" },
    },
    {
      id: "modules/vg",
      data: {
        name: "VG",
        summary: "<p>Toolkit to manipulate and analyze graphical genomes, including read alignment.</p>",
      },
    },
    {
      id: "modules/whatshap",
      data: {
        name: "WhatsHap",
        summary: "<p>Phasing genomic variants using DNA reads (aka read-based phasing, or haplotype assembly).</p>",
      },
    },
    {
      id: "modules/xengsort",
      data: { name: "Xengsort", summary: "<p>Fast xenograft read sorter based on space-efficient k-mer hashing.</p>" },
    },
    {
      id: "modules/xenium",
      data: {
        name: "Xenium",
        summary: "<p>Spatial transcriptomics platform from 10x Genomics that provides subcellular resolution.</p>",
      },
    },
    { id: "modules/xenome", data: { name: "Xenome", summary: "<p>Classifies reads from xenograft sources.</p>" } },
  ]}
/>
